### PR TITLE
Support tracing Catchpoint requests

### PR DESCRIPTION
**Pull request by erdogan-yesil**

### CODE REVIEW

File: thundra/wrappers/cp_wrapper_utils.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    except Exception as e:
+        logger.error('ERROR: {}'.format(e))
+
+
+def is_triggered_from_catchpoint(request):
+    if request and request.headers:
+        headers = request.headers
+        user_agent = headers.get(constants.HTTPHeaders['USER_AGENT'])
+        return 'Catchpoint' in user_agent if user_agent else False
+    return True
```
Comment (serkan-ozal) [inline thundra/wrappers/cp_wrapper_utils.py:28]: `return False`

File: thundra/wrappers/cp_wrapper_utils.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+def create_catchpoint_request_invocation(execution_context, application_info, region_name, country_name, city_name,
+                                         test_id, trace_id, transaction_id, span_id, start_timestamp, finish_timestamp,
+                                         resource):
+    if not execution_context.transaction_id:
+        execution_context.transaction_id = str(uuid.uuid4())
+    invocation_data = {
+        'id': str(uuid.uuid4()),
+        'type': 'Invocation',
+        'agentVersion': constants.THUNDRA_AGENT_VERSION,
```
Comment (serkan-ozal) [inline thundra/wrappers/cp_wrapper_utils.py:72]: `CatchpointProperties.get('AGENT_VERSION')`

File: thundra/wrappers/cp_wrapper_utils.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    return invocation_data
+
+
+def create_catchpoint_request_span(application_info, root_span, resource, region_name, country_name,
+                                   city_name, test_id, trace_id, transaction_id, span_id, start_timestamp,
+                                   finish_timestamp):
+        span_data = {
+            'id': span_id,
+            'type': 'Span',
+            'agentVersion': constants.THUNDRA_AGENT_VERSION,
```
Comment (serkan-ozal) [inline thundra/wrappers/cp_wrapper_utils.py:114]: `CatchpointProperties.get('AGENT_VERSION')`

File: thundra/wrappers/cp_wrapper_utils.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+                constants.CatchpointTags.get('TEST_ID'): test_id
+            }
+        }
+        inject_application_info(span_data, application_info)
+        return span_data
+
+
+def inject_application_info(data, application_info):
+    data.update(application_info)
+    data['applicationRuntime'] = ApplicationInfoProvider.APPLICATION_RUNTIME
```
Comment (serkan-ozal) [inline thundra/wrappers/cp_wrapper_utils.py:145]: `applicationRuntime` should be `None`

File: thundra/wrappers/cp_wrapper_utils.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+                                                           country_name, city_name, test_id, trace_id, transaction_id,
+                                                           span_id, start_timestamp, finish_timestamp, resource)
+    execution_context.report(create_report_data(api_key, 'Invocation', invocation_data))
+
+    # create catchpoint span data
+    span_data = create_catchpoint_request_span(application_info, span, resource, region_name, country_name, city_name, 
+                                               test_id, trace_id, transaction_id, span_id, start_timestamp, finish_timestamp)
+    execution_context.report(create_report_data(api_key, 'Span', span_data))
+
+    if response and hasattr(response, 'headers'):
```
Comment (serkan-ozal) [inline thundra/wrappers/cp_wrapper_utils.py:60]: Even `response` has no headers, you need initialize headers with empty dictionary and inject `constants.THUNDRA_TRACE_ID_KEY` into headers

File: thundra/wrappers/cp_wrapper_utils.py (hunk 2 above)
Comment (erdogan-yesil) [reply thundra/wrappers/cp_wrapper_utils.py:72]: done

File: thundra/wrappers/cp_wrapper_utils.py (hunk 3 above)
Comment (erdogan-yesil) [reply thundra/wrappers/cp_wrapper_utils.py:114]: done

File: thundra/wrappers/cp_wrapper_utils.py (hunk 5 above)
Comment (erdogan-yesil) [reply thundra/wrappers/cp_wrapper_utils.py:60]: done

File: thundra/wrappers/cp_wrapper_utils.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+def generate_catchpoint_application_name(region_name, country_name, city_name):
+    if city_name:
+        return city_name
+    elif country_name:
+        return country_name
+    elif region_name:
+        return region_name
+    else:
+        return constants.DEFAULT_APPLICATION_NAME
```
[the remark this answers is not in the document]
Comment (erdogan-yesil) [reply thundra/wrappers/cp_wrapper_utils.py:196]: done

File: thundra/wrappers/cp_wrapper_utils.py (hunk 1 above)
Comment (erdogan-yesil) [reply thundra/wrappers/cp_wrapper_utils.py:28]: done

File: thundra/wrappers/cp_wrapper_utils.py (hunk 4 above)
Comment (erdogan-yesil) [reply thundra/wrappers/cp_wrapper_utils.py:145]: applicationRuntime should not be null

File: thundra/wrappers/cp_wrapper_utils.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+            }
+        }
+        inject_application_info(span_data, application_info)
+        return span_data
+
+
+def inject_application_info(data, application_info):
+    data.update(application_info)
+    data['applicationRuntime'] = ApplicationInfoProvider.APPLICATION_RUNTIME
+    data['applicationRuntimeVersion'] = ApplicationInfoProvider.APPLICATION_RUNTIME_VERSION
```
[the remark this answers is not in the document]
Comment (erdogan-yesil) [reply thundra/wrappers/cp_wrapper_utils.py:146]: applicationRuntimeVersion should not be null

**Pull request merged**